examples/json.dx

```diff
@@ -0,0 +1,258 @@
+
+
+' # Declarative Graphing
+
+
+' ## JSON Implementation
+
+def join (extra: List a) (lists:n=>(List a)) : List a =
+    concat $ for i.
+        case ordinal i == (size n - 1) of
+             True -> lists.i
+             False -> lists.i <> extra
+
+
+data JValue = AsJValue String
+-- TODO - once Dex supports recursive ADT JValue becomes Value.
+
+data Value = 
+     AsObject (List (String & JValue))
+     AsArray (List JValue)
+     AsString String
+     AsFloat Float
+     AsInt Int
+
+interface ToJSON a
+  toJSON : a -> Value
+
+
+-- These three are private methods. Users should use type classes.
+def escape (x:JValue) : String =
+    (AsJValue y) = x
+    y
+
+
+def collapse (x:Value) : JValue =
+    AsJValue $ case x of
+         AsString y ->  "\"" <> y <> "\""
+         AsFloat y -> show y
+         AsInt y -> show y
+         AsObject (AsList _ y) ->
+              ("{" <> (join ", " $ for i.
+                 (k, v) = y.i 
+                 "\"" <> k <> "\"" <> ":" <> (escape v)) <> "}")
+         AsArray (AsList _ y) -> ("[" <> (join ", " $ for i. escape y.i) <> "]")
+
+def hide [ToJSON a] (x:a) : JValue =
+    collapse $ toJSON x
+
+instance Show Value 
+  show = \x. escape $ collapse x
+
+instance ToJSON String
+  toJSON = AsString
+  
+instance ToJSON Int
+  toJSON = AsInt
+  
+instance ToJSON Float
+  toJSON = AsFloat
+  
+instance ToJSON Value
+  toJSON = id
+  
+instance [ToJSON v] ToJSON ((Fin n) => v)
+  toJSON = \x . AsArray $ AsList _ $ for i. hide x.i
+
+instance [ToJSON v] ToJSON ((Fin n) => (String & v))
+  toJSON = \x . AsObject $ AsList _ $ for i. (fst x.i, hide $ snd x.i)
+
+instance [ToJSON v] ToJSON (List (String & v))
+  toJSON = \(AsList _ x) . toJSON x
+
+
+' ## Graph Grammars (Vega-Lite Spec)
+
+Options = List (String & String)
+
+data Mark =
+     AsMark String Options
+
+data EncodingType =
+     Quantitative
+     Nominal
+     Ordinal
+
+instance Show EncodingType
+         show = (\ x.
+              case x of
+                   Quantitative -> "quantitative"
+                   Nominal -> "nominal"
+                   Ordinal -> "ordinal")
+                   
+data Channel =
+     Y
+     X
+     Color
+     Tooltip
+     HREF
+     Row
+     Col
+     Size
+     
+instance Show Channel
+         show = (\ x.
+              case x of
+                   Y -> "y"
+                   X -> "x"
+                   Color -> "color"
+                   Tooltip -> "tooltip"
+                   HREF -> "href"
+                   Size -> "size"
+                   Row -> "row"
+                   Col -> "col")
+
+data Encoding key =
+     AsEncoding Channel key EncodingType Options
```

Comment:
How about instead of putting the key as part of `Encoding`, we just require the list of encodings to be sorted according to the list of fields of the data frame?

Reply:
Changed to this style.